### PR TITLE
Support ISBN citations

### LINKED ISSUE #14: ISBN citation of books
While many scholarly books have DOIs, there are lot's of books out there with just ISBNs. As a standardized identifier, they seem appropriate for Manubot book citations. We discussed the issue in https://github.com/greenelab/deep-review/issues/387 and found serveral ISBN lookup services.

However, none of them are aware of the [2012 Open Access Book](https://mitpress.mit.edu/books/open-access) indicating poor coverage:

+ https://isbnsearch.org/isbn/9780262302524
+ http://isbndb.com/search/all?query=9780262302524
+ http://xisbn.worldcat.org/webservices/xid/isbn/9780262302524?method=getMetadata&format=xml&fl=*

**Update**: I think `9780262302524` is the wrong ISBN, but was listed on the MIT Press website.

@agapow commented:

> My impression of worldcat and isbndb is that there's quite a lag before titles get listed (and they may never get listed). Not a solution, but we probably need another way.

So the purpose of this issue is to note that ISBN citation would be valuable and to track any potential solutions.

### PR DESCRIPTION
**Pull request by dhimmel**
Work in progress

Closes #14

### CODE REVIEW

**Comment by dhimmel** [comment]
I ran `manubot cite` with all 500 test DOIs from [here](https://gist.github.com/adrianshort/1253715/998fb77ec926d8b3ed22819f1d7a0ebe29569516) with codebase https://github.com/greenelab/manubot/pull/79/commits/cde55ca734514c49a9114c51d7e1b7f64974553a.

Got the following output:

<details>
```
## WARNING
Error parsing issued date for ISBN 9780006499152.
Metadata retrieved by get_isbn_citeproc_citoid from https://en.wikipedia.org/api/rest_v1/data/citation/mediawiki/9780006499152
The following value was returned for the date field:
(1996 [printing])
## WARNING
Error parsing issued date for ISBN 9780091874568.
Metadata retrieved by get_isbn_citeproc_citoid from https://en.wikipedia.org/api/rest_v1/data/citation/mediawiki/9780091874568
The following value was returned for the date field:
(2001 [printing])
## WARNING
Error parsing issued date for ISBN 9780099283577.
Metadata retrieved by get_isbn_citeproc_citoid from https://en.wikipedia.org/api/rest_v1/data/citation/mediawiki/9780099283577
The following value was returned for the date field:
2001, ©1977
## WARNING
Error parsing issued date for ISBN 9780099512790.
Metadata retrieved by get_isbn_citeproc_citoid from https://en.wikipedia.org/api/rest_v1/data/citation/mediawiki/9780099512790
The following value was returned for the date field:
2008, ©2007
## WARNING
Error parsing issued date for ISBN 9780099549024.
Metadata retrieved by get_isbn_citeproc_citoid from https://en.wikipedia.org/api/rest_v1/data/citation/mediawiki/9780099549024
The following value was returned for the date field:
2011, ©2010
## WARNING
Error parsing issued date for ISBN 9780140052473.
Metadata retrieved by get_isbn_citeproc_citoid from https://en.wikipedia.org/api/rest_v1/data/citation/mediawiki/9780140052473
The following value was returned for the date field:
1980, ©1969
## WARNING
Error parsing issued date for ISBN 9780140069990.
Metadata retrieved by get_isbn_citeproc_citoid from https://en.wikipedia.org/api/rest_v1/data/citation/mediawiki/9780140069990
The following value was returned for the date field:
1985, ©1978
## WARNING
Error parsing issued date for ISBN 9780140135824.
Metadata retrieved by get_isbn_citeproc_citoid from https://en.wikipedia.org/api/rest_v1/data/citation/mediawiki/9780140135824
The following value was returned for the date field:
1990, ©1968
## WARNING
Error parsing issued date for ISBN 9780140367980.
Metadata retrieved by get_isbn_citeproc_citoid from https://en.wikipedia.org/api/rest_v1/data/citation/mediawiki/9780140367980
The following value was returned for the date field:
(1994 [printing])
## WARNING
Error parsing issued date for ISBN 9780141014593.
Metadata retrieved by get_isbn_citeproc_citoid from https://en.wikipedia.org/api/rest_v1/data/citation/mediawiki/9780141014593
The following value was returned for the date field:
2006, ©2005
## WARNING
Error parsing issued date for ISBN 9780307398512.
Metadata retrieved by get_isbn_citeproc_citoid from https://en.wikipedia.org/api/rest_v1/data/citation/mediawiki/9780307398512
The following value was returned for the date field:
2011, c2010
## WARNING
Error parsing issued date for ISBN 9780330309578.
Metadata retrieved by get_isbn_citeproc_citoid from https://en.wikipedia.org/api/rest_v1/data/citation/mediawiki/9780330309578
The following value was returned for the date field:
1989, ©1988
## WARNING
Error parsing issued date for ISBN 9780330313100.
Metadata retrieved by get_isbn_citeproc_citoid from https://en.wikipedia.org/api/rest_v1/data/citation/mediawiki/9780330313100
The following value was returned for the date field:
1990, ©1989
## WARNING
Error parsing issued date for ISBN 9780330322690.
Metadata retrieved by get_isbn_citeproc_citoid from https://en.wikipedia.org/api/rest_v1/data/citation/mediawiki/9780330322690
The following value was returned for the date field:
1992, ©1991
## WARNING
Error parsing issued date for ISBN 9780393301663.
Metadata retrieved by get_isbn_citeproc_citoid from https://en.wikipedia.org/api/rest_v1/data/citation/mediawiki/9780393301663
The following value was returned for the date field:
1984, ©1967
## WARNING
Error parsing issued date for ISBN 9780394729114.
Metadata retrieved by get_isbn_citeproc_citoid from https://en.wikipedia.org/api/rest_v1/data/citation/mediawiki/9780394729114
The following value was returned for the date field:
1985, ©1984
## WARNING
Error parsing issued date for ISBN 9780552144957.
Metadata retrieved by get_isbn_citeproc_citoid from https://en.wikipedia.org/api/rest_v1/data/citation/mediawiki/9780552144957
The following value was returned for the date field:
1999, ©1998
## WARNING
Error parsing issued date for ISBN 9780552772815.
Metadata retrieved by get_isbn_citeproc_citoid from https://en.wikipedia.org/api/rest_v1/data/citation/mediawiki/9780552772815
The following value was returned for the date field:
2007, ©2006
## WARNING
Error parsing issued date for ISBN 9780552773317.
Metadata retrieved by get_isbn_citeproc_citoid from https://en.wikipedia.org/api/rest_v1/data/citation/mediawiki/9780552773317
The following value was returned for the date field:
2007, ©2006
## WARNING
Error parsing issued date for ISBN 9780552773898.
Metadata retrieved by get_isbn_citeproc_citoid from https://en.wikipedia.org/api/rest_v1/data/citation/mediawiki/9780552773898
The following value was returned for the date field:
2007, ©2005
## WARNING
Error parsing issued date for ISBN 9780552998482.
Metadata retrieved by get_isbn_citeproc_citoid from https://en.wikipedia.org/api/rest_v1/data/citation/mediawiki/9780552998482
The following value was returned for the date field:
2000, ©1999
## WARNING
Error parsing issued date for ISBN 9780719561023.
Metadata retrieved by get_isbn_citeproc_citoid from https://en.wikipedia.org/api/rest_v1/data/citation/mediawiki/9780719561023
The following value was returned for the date field:
(2004 printing)
## WARNING
Error parsing issued date for ISBN 9780747580799.
Metadata retrieved by get_isbn_citeproc_citoid from https://en.wikipedia.org/api/rest_v1/data/citation/mediawiki/9780747580799
The following value was returned for the date field:
(2006 printing)
## WARNING
Error parsing issued date for ISBN 9780749320201.
Metadata retrieved by get_isbn_citeproc_citoid from https://en.wikipedia.org/api/rest_v1/data/citation/mediawiki/9780749320201
The following value was returned for the date field:
1996, ©1995
## WARNING
Error parsing issued date for ISBN 9780800626792.
Metadata retrieved by get_isbn_citeproc_citoid from https://en.wikipedia.org/api/rest_v1/data/citation/mediawiki/9780800626792
The following value was returned for the date field:
1992-<2013>
## WARNING
Error parsing issued date for ISBN 9780954642815.
Metadata retrieved by get_isbn_citeproc_citoid from https://en.wikipedia.org/api/rest_v1/data/citation/mediawiki/9780954642815
The following value was returned for the date field:
(2005 [printing])
## WARNING
Error parsing issued date for ISBN 9780957921849.
Metadata retrieved by get_isbn_citeproc_citoid from https://en.wikipedia.org/api/rest_v1/data/citation/mediawiki/9780957921849
The following value was returned for the date field:
2004, ©2003
## WARNING
Error parsing issued date for ISBN 9781401359287.
Metadata retrieved by get_isbn_citeproc_citoid from https://en.wikipedia.org/api/rest_v1/data/citation/mediawiki/9781401359287
The following value was returned for the date field:
[2004?], ©2002
## WARNING
Error parsing issued date for ISBN 9781405208093.
Metadata retrieved by get_isbn_citeproc_citoid from https://en.wikipedia.org/api/rest_v1/data/citation/mediawiki/9781405208093
The following value was returned for the date field:
2002, ©1968
## WARNING
Error parsing issued date for ISBN 9781406306804.
Metadata retrieved by get_isbn_citeproc_citoid from https://en.wikipedia.org/api/rest_v1/data/citation/mediawiki/9781406306804
The following value was returned for the date field:
2007, ©1997
## WARNING
Error parsing issued date for ISBN 9781408808870.
Metadata retrieved by get_isbn_citeproc_citoid from https://en.wikipedia.org/api/rest_v1/data/citation/mediawiki/9781408808870
The following value was returned for the date field:
2010, ©2010
## ERROR
too many values to unpack (expected 1)
## WARNING
Error parsing issued date for ISBN 9781843540236.
Metadata retrieved by get_isbn_citeproc_citoid from https://en.wikipedia.org/api/rest_v1/data/citation/mediawiki/9781843540236
The following value was returned for the date field:
(2005 printing)
## WARNING
Error parsing issued date for ISBN 9781846941764.
Metadata retrieved by get_isbn_citeproc_citoid from https://en.wikipedia.org/api/rest_v1/data/citation/mediawiki/9781846941764
The following value was returned for the date field:
c 2008
## WARNING
Error parsing issued date for ISBN 9781847670359.
Metadata retrieved by get_isbn_citeproc_citoid from https://en.wikipedia.org/api/rest_v1/data/citation/mediawiki/9781847670359
The following value was returned for the date field:
2007, ©2006
## CRITICAL
Failure: exiting with code 1 due to logged errors
```

</details>

So plenty of DOIs with peculiar date fields. Will need to revise date parsing method. Also one error `too many values to unpack (expected 1)` that I will have to track down.

**Comment by dhimmel** [comment]
BTW builds are failing because the NCBI Citation Exporter is down. See https://twitter.com/dhimmel/status/1061787168820092929

**Comment by agitter** [comment]
I can review but will need a day or two to get to it.

File: manubot/cite/isbn.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
+import collections
+import json
+import logging
+import re
+
+
+def get_isbn_citeproc(isbn):
```
Comment (agitter) [inline manubot/cite/isbn.py]: Are you allowing flexibility to later query multiple ISBN metadata sources?

File: manubot/cite/isbn.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    return get_isbn_citeproc_citoid(isbn)
+
+
+def get_isbn_citeproc_isbnlib(isbn):
```
Comment (agitter) [inline manubot/cite/isbn.py]: It looks like this function isn't used currently.  Will it be later?

File: manubot/cite/isbn.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    Return CSL JSON Data for an ISBN using the Wikipedia Citoid API.
+    https://en.wikipedia.org/api/rest_v1/#!/Citation/getCitation
+    """
+    import isbnlib
+    import requests
+    isbn = isbnlib.to_isbn13(isbn)
+    url = f'https://en.wikipedia.org/api/rest_v1/data/citation/mediawiki/{isbn}'
+    response = requests.get(url)
+    result = response.json()
+    if isinstance(result, dict):
```
Comment (agitter) [inline manubot/cite/isbn.py]: What type of object is `result` here if it is not a `dict`?

File: tests/test_citations.py [hunk 4]
```diff
@@ -245,6 +254,39 @@ def test_citation_to_citeproc_pubmed_book():
         citation_to_citeproc('pmid:29227604')
 
 
+def test_citation_to_citeproc_isbn():
+    csl_item = citation_to_citeproc('isbn:9780387950693')
+    assert csl_item['type'] == 'book'
+    assert csl_item['title'] == 'Complex Analysis'
```
Comment (agitter) [inline tests/test_citations.py]:
```suggestion
    assert csl_item['title'] == 'Complex analysis'
```

File: tests/test_citations.py [hunk 5]
```diff
@@ -245,6 +254,39 @@ def test_citation_to_citeproc_pubmed_book():
...
+
+
+def test_citation_to_citeproc_isbn_title_with_quotation_mark():
+    csl_item = citation_to_citeproc('isbn:9780312353780')
+    assert csl_item['type'] == 'book'
+    assert csl_item['title'].startswith('"F" is for Fugitive')
```
Comment (agitter) [inline tests/test_citations.py]: This test case is also failing.  Does it need to be `\"F\" is for fugitive`?

File: tests/test_citations.py [hunk 6]
```diff
@@ -245,6 +254,39 @@ def test_citation_to_citeproc_pubmed_book():
...
+
+
+def test_get_isbn_citeproc_citoid_weird_date():
+    """
+    isbn:9780719561023 has a date value of "(2004 printing)"
+    https://en.wikipedia.org/api/rest_v1/data/citation/mediawiki/9780719561023
```
Comment (agitter) [inline tests/test_citations.py]: Note that the author is also weird for this book.  Manubot produces:
```
    "author": [
      {
        "given": "1969-",
        "family": "Frey, James,"
      }
    ],
```

File: manubot/cite/isbn.py (hunk 3 above)
Comment (dhimmel) [reply manubot/cite/isbn.py]: If things go well, result is a list of one element.

File: manubot/cite/isbn.py (hunk 2 above)
Comment (dhimmel) [reply manubot/cite/isbn.py]: This touches on a larger question: do we retain metadata methods that are not used as the default method for a certain metadata source. I think we should because it allows:

1. us to easily switch to the other source in the future
2. may be helpful for certain users who are using manubot for citation metadata retreival

So I kept this in here because `isbnlib` is now a dependency, and thought it made sense to at least make it's metadata retrieval an option.

File: manubot/cite/isbn.py (hunk 1 above)
Comment (dhimmel) [reply manubot/cite/isbn.py]: At some point, we may want to have this function fallback to alternative sources.

In addition, this makes it easy to switch whichever function is the default.

File: tests/test_citations.py (hunk 6 above)
Comment (dhimmel) [reply tests/test_citations.py]: Yes authors are quite bad. In almost all cases I tried with Citoid, the author metadata was imperfect.

File: tests/test_citations.py (hunk 5 above)
Comment (dhimmel) [reply tests/test_citations.py]: I meant to be testing `get_isbn_citeproc_isbnlib`, which does produce the correct title.

**Comment by dhimmel** [comment]
> Given all the corner cases, ISBN support will likely be similar to URL support with Greycite. Sometimes it will work. Other times it will seed the CSL JSON for manual corrections.

Yes! It will be nice however to start using ISBNs to identify books rather than URLs.

File: manubot/cite/isbn.py (hunk 2 above)
Comment (agitter) [reply manubot/cite/isbn.py]: Your reasoning makes sense to me.  I do think it would improve readability to have a comment that `get_isbn_citeproc_isbnlib` is an alternative ISBN metadata retrieval function that is not currently used in the standard ISBN processing pipeline.

Comment (agitter) [reply manubot/cite/isbn.py]: By the way, I resolved the other parts of our discussion that seem to be addressed.  This suggestion is the only remaining comment that may be actionable, depending on what you think.

Comment (dhimmel) [reply manubot/cite/isbn.py]: How about https://github.com/greenelab/manubot/pull/79/commits/9337ef754c3738a0a40cb452ea653aaca56d8a4f?

Comment (agitter) [reply manubot/cite/isbn.py]: LGTM
